gitlab/client_repository_tree.go

```diff
@@ -0,0 +1,50 @@
+/*
+Copyright 2020 The Flux CD contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gitlab
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/fluxcd/go-git-providers/gitprovider"
+)
+
+// TreeClient implements the gitprovider.TreeClient interface.
+var _ gitprovider.TreeClient = &TreeClient{}
+
+// TreeClient operates on the trees in a specific repository.
+type TreeClient struct {
+	*clientContext
+	ref gitprovider.RepositoryRef
+}
+
+// Create creates,updates,deletes a tree
+func (c *TreeClient) Create(ctx context.Context, tree *gitprovider.TreeInfo) (*gitprovider.TreeInfo, error) {
+	return nil, fmt.Errorf("error creaing tree %s. not implemented in gitlab yet", tree.SHA)
+
+}
+
+// Get returns a tree
+func (c *TreeClient) Get(ctx context.Context, sha string, recursive bool) (*gitprovider.TreeInfo, error) {
+	return nil, fmt.Errorf("error getting tree %s. not implemented in gitlab yet", sha)
+
+}
+
+// List files (blob) in a tree
+func (c *TreeClient) List(ctx context.Context, sha string, recursive bool) ([]*gitprovider.TreeEntry, error) {
+	return nil, fmt.Errorf("error listing tree items %s. not implemented in gitlab yet", sha)
```

Comment:
GitLab forces pagination on this API and limits to 100 per_page, however it does allow you to start at a *subpath* of the repo: https://docs.gitlab.com/ee/api/repositories.html

**Proposal**

In [other places in this repository we auto-paginate](https://github.com/fluxcd/go-git-providers/blob/main/gitlab/gitlabclient.go#L240) results so we could follow that pattern here too. 

**Notes**

- For our particular case of deleting all the files in a subpath that should run on a cluster, the worst case could be 10000+ files, which is 100 pagination requests and will probably timeout. Asking the user to manually remove those files would be the workaround and seems reasonable. In the future we could add an `...Options` param to this API to allow more control over pagination for other users of this library with other use cases.
- I would _guess_ the average case would be more on the order of < 100 files which we could handle in a single request or two.
- As Gitlab does not have a tree api for doing cleverer manipulations, the only other options to perform recursive deletion operations efficiently for gitlab seems to be:
  1. [Grab file archive](https://docs.gitlab.com/ee/api/repositories.html#get-file-archive) (limited to 5 requests / minute on gitlab.com) 
  2. Git clone.

Reply:
> In [other places in this repository we auto-paginate](https://github.com/fluxcd/go-git-providers/blob/main/gitlab/gitlabclient.go#L240) results so we could follow that pattern here too.

This is the main reason I recommend against using go-git-providers, it has no control over pages being loaded.

This pattern is probably a DoS risk, because you can't control how many pages might be loaded for any resource.

Reply:
- with the current limitation mentioned in [this issue](https://github.com/fluxcd/go-git-providers/issues/158) and sticking to the limit of 100 files when gitlab and giving a warning of this limitation so that if there are more files in the specific directory they are to be deleted manually, is this a possible temporary solution?
- how about the number of pages loaded for a resource to be limited with a max_pages flag passed